### PR TITLE
Precomputation of primes

### PR DESCRIPTION
**Pull request by tilenmarc**
This PR introduces the following:
- Each scheme that depends on Z_p group now includes a new function `NewSchemePrecomp` that chooses the Z_p group based on precomputed prime numbers and generators. This allows a fast scheme initialization for realistic parameters, i.e. prime numbers with bit length 2048 or more.
- In each test of a scheme that depends on Z_p group `NewSchemePrecomp` is used.
- Additionally, automatic code review by service Codacy is introduced. Hence all the suggested small code changes are done, so that the library has no issues any more.
- A badge is added in `readme.md` file that shows a grade for the code quality given by Codacy

### CODE REVIEW

File: innerprod/fullysec/damgard_test.go [hunk 1]
```diff
@@ -26,14 +26,24 @@ import (
...
-	modulusLength := 64
-
-	damgard, err := fullysec.NewDamgard(l, modulusLength, bound)
+type damgardTestParam struct {
+	name string
+	modulusLength int
+	precomputed bool
+}
 
+func testFullySec_DamgardDDHFromParam(t *testing.T, param damgardTestParam) {
```
Comment (tilenmarc) [inline innerprod/fullysec/damgard_test.go]: Codacy Issue found: [don't use underscores in Go names; func testFullySec_DamgardDDHFromParam should be testFullySecDamgardDDHFromParam](https://app.codacy.com/app/tilenmarc/gofe_2/pullRequest?prid=4432579)

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 34
- Added 5
           

Complexity increasing per file
==============================
- innerprod/fullysec/damgard_test.go  1
- innerprod/fullysec/damgard_multi_test.go  1
- innerprod/fullysec/damgard.go  1
- innerprod/fullysec/damgard_dec_multi_test.go  1
- innerprod/simple/ddh.go  4
- innerprod/simple/ddh_multi_test.go  1
- innerprod/simple/ddh_test.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/gofe_2/gofe_2/pullRequest?prid=4432579&bid=14999526)

File: innerprod/simple/ddh_test.go [hunk 2]
```diff
@@ -26,14 +26,24 @@ import (
...
 )
 
-func TestSimple_DDH(t *testing.T) {
+type dDHTestParam struct {
+	name string
+	modulusLength int
+	precomputed bool
+}
+
+func testSimple_DDHFromParam(t *testing.T, param dDHTestParam) {
```
Comment (tilenmarc) [inline innerprod/simple/ddh_test.go]: Codacy Issue found: [don't use underscores in Go names; func testSimple_DDHFromParam should be testSimpleDDHFromParam](https://app.codacy.com/app/tilenmarc/gofe_2/pullRequest?prid=4432579)

File: innerprod/simple/ddh_multi_test.go [hunk 3]
```diff
@@ -26,20 +26,30 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestSimple_MultiDDH(t *testing.T) {
+func testSimple_MultiDDHFromParam(t *testing.T, param dDHTestParam) {
```
Comment (tilenmarc) [inline innerprod/simple/ddh_multi_test.go]: Codacy Issue found: [don't use underscores in Go names; func testSimple_MultiDDHFromParam should be testSimpleMultiDDHFromParam](https://app.codacy.com/app/tilenmarc/gofe_2/pullRequest?prid=4432579)

File: innerprod/fullysec/damgard_dec_multi_test.go [hunk 4]
```diff
@@ -26,18 +26,23 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestFullySec_DamgardDecMulti(t *testing.T) {
+func testFullySec_DamgardDecMultiFromParam(t *testing.T, param damgardTestParam) {
```
Comment (tilenmarc) [inline innerprod/fullysec/damgard_dec_multi_test.go]: Codacy Issue found: [don't use underscores in Go names; func testFullySec_DamgardDecMultiFromParam should be testFullySecDamgardDecMultiFromParam](https://app.codacy.com/app/tilenmarc/gofe_2/pullRequest?prid=4432579)

File: innerprod/fullysec/damgard_multi_test.go [hunk 5]
```diff
@@ -26,17 +26,23 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestFullySec_DamgardMultiDDH(t *testing.T) {
+func testFullySec_DamgardMultiDDHFromParam(t *testing.T, param damgardTestParam) {
```
Comment (tilenmarc) [inline innerprod/fullysec/damgard_multi_test.go]: Codacy Issue found: [don't use underscores in Go names; func testFullySec_DamgardMultiDDHFromParam should be testFullySecDamgardMultiDDHFromParam](https://app.codacy.com/app/tilenmarc/gofe_2/pullRequest?prid=4432579)

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 34
           

Complexity increasing per file
==============================
- innerprod/fullysec/damgard_test.go  1
- innerprod/simple/ddh_multi_test.go  1
- innerprod/simple/ddh_test.go  1
- innerprod/fullysec/damgard.go  1
- innerprod/fullysec/damgard_multi_test.go  1
- innerprod/fullysec/damgard_dec_multi_test.go  1
- innerprod/simple/ddh.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/gofe_2/gofe_2/pullRequest?prid=4432579&bid=14999526)

**Comment by tilenmarc** [comment]
I spit the test to random and precomputed primes. As you will see, there is a nice way to do so in Go. I added a bit more comments to the code.

**Comment by miha-stopar** [comment]
Nice one! I only have one additional minor comment - dDHTestParam looks somehow strange, perhaps it would be better having ddhTestParam?

**Comment by tilenmarc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 34
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/gofe_2/gofe_2/pullRequest?prid=4432579&bid=14999526)

**Pull request merged**